### ISSUE TITLE
Point cloud shows up significantly above or below terrain surface

### ISSUE THREAD

**Issue by iannesbitt** [PermafrostDiscoveryGateway/viz-points#14]
In certain areas, point clouds are not well aligned in Z with whatever terrain model is being used. The point cloud ground surface ends up meters above or below the draped ground surface in Cesium but located ok in X and Y. Having points above the ground surface wouldn't be as much of a problem if the ground surface weren't such a vital part of the navigation mechanics of the Cesium js viewer. Having points below ground fully hides them from view of the user.

This happens with all coordinate systems tried so far, whether reprojected in PDAL, reprojected in LAStools, or untransformed and projected as-is in Cesium. The only solution so far is a manual transform of Z in LAStools, which is unfeasible on a large scale.

Guiding questions so far:
- Is this a Cesium problem? Am I missing the geoid or ellipsoid or something?
- Is this a projection issue? Probably not, considering multiple projections tested
- Is this a positioning issue? Probably not, considering this is an affliction with official datasets as well
- ?
- **Where in relation to the ground do we want 3D data to actually show up in the viewer?**

Below are examples.

Seen here is the Lake Placid test dataset (derived from an officially released [USGS dataset](https://www.sciencebase.gov/catalog/item/5bb4b4e3e4b08583a5d8d383)) a few tens of meters above the NED-derived ground surface:
![Lake Placid test dataset](https://user-images.githubusercontent.com/18689918/233394881-84211d92-c514-44c8-9815-02601bce4752.png)

Below, an example of terrain off and terrain on in which the road surface from a sample dataset in Nome, AK ends up beneath both the NED and ArcticDEM terrain models when terrain is enabled:
![terrain off](https://user-images.githubusercontent.com/18689918/233395217-6a1af35d-5f38-4f88-b6af-ce71cffe99c6.png)
![terrain on](https://user-images.githubusercontent.com/18689918/233395261-6719c0a9-150d-484d-93af-f75b006836e2.png)

Finally, a dataset from Huntsville, AL hovers 10 meters or more above the NED model:
![image](https://user-images.githubusercontent.com/18689918/233396854-0e0ed14f-c2d4-4e4d-86f0-75e08ff5f6c5.png)

**Comment by iannesbitt** [comment]
Temporary solution is to adjust per-region using an adjustment factor. Here is Nome, AK data with +9 m Z translation: 
![image](https://user-images.githubusercontent.com/18689918/236521080-37bf5c76-b758-4a93-897e-b2764519647f.png)

**Comment by iannesbitt** [comment]
A solution to draw points that exist below the terrain:
```javascript
viewer.scene.globe.depthTestAgainstTerrain = false;
```

**Comment by iannesbitt** [comment]
Adjusting the Z values to accommodate the [geoid height](https://geodesy.noaa.gov/api/geoid/ght?lat=44.25&lon=-73.96) seems to remedy the issue at least for the test dataset in Lake Placid:
![image](https://github.com/PermafrostDiscoveryGateway/viz-points/assets/18689918/f9d5969c-d9fb-432b-bb75-538acf97e028)

**Comment by iannesbitt** [comment]
Geoid height API from NOAA: https://geodesy.noaa.gov/web_services/geoid.shtml

Usage example for CONUS: https://geodesy.noaa.gov/api/geoid/ght?lat=44.257&lon=-73.965
for Alaska: https://geodesy.noaa.gov/api/geoid/ght?lat=64.506&lon=-165.399&model=12

**Comment by iannesbitt** [comment]
Here is a visual representation of what I'm doing to correct the geoid height (in the LAS files) to ellipsoid height (used by Cesium's terrain model). Cesium uses `h` to define its terrain whereas the LAS datasets are using `H`. The difference between the ellipsoid and geoid is `N`. When I add `H+N` I get something that much more closely resembles the Cesium topography. 
![image](https://github.com/PermafrostDiscoveryGateway/viz-points/assets/18689918/5b5dd0df-e2fb-4ea9-90e7-575287a069e6)
I think it may be worthwhile to add in a geoid height adjustment feature that uses the NOAA API to get `N` from one of their models and applies it as a Z translation to approximate ellipsoid height. I still need to try to understand if there's a way to tell if a dataset has its height values based on the geoid or the ellipsoid, but the adjustment shouldn't be hard to code.

**Comment by iannesbitt** [comment]
This service may be of use if NAVD88 is the vertical datum we have to correct for: https://www.ngs.noaa.gov/web_services/ncat/lat-long-height-service.shtml

**Comment by iannesbitt** [comment]
Emmanuel has confirmed that the flight line elevation values were converted to geoid height (EGM2008) during an early stage of post-processing. I'm not sure LAStools has any built-in capability to convert from EGM height to ellipsoid height, so adjustment built into the pipeline may have to be from third party software (perhaps [pygeodesy](https://stackoverflow.com/a/68138068/4648080)). Luckily for most datasets I think the adjustment value can be applied as a fixed value (i.e. `N` can be computed and applied per-file and not per-point)

At the moment the way QT Modeler attaches geoid information to LAS files is by using compound coordinate system well-known text strings. An example is given below. Currently `viz-points` is unable to parse COMPD_CS WKT strings, as it just grabs the last EPSG code from the string, rather than the code from the `AUTHORITY` field of the `PROJCS`. I will make an issue to deal with this shortly.

```
COMPD_CS["WGS 84 / UTM zone 3N + EGM2008 height",PROJCS["WGS 84 / UTM zone 3N",GEOGCS["WGS 84",DATUM["WGS_1984",SPHEROID["WGS 84",6378137,298.257223563,AUTHORITY["EPSG","7030"]],AUTHORITY["EPSG","6326"]],PRIMEM["Greenwich",0,AUTHORITY["EPSG","8901"]],UNIT["degree",0.0174532925199433,AUTHORITY["EPSG","9122"]],AUTHORITY["EPSG","4326"]],PROJECTION["Transverse_Mercator"],PARAMETER["latitude_of_origin",0],PARAMETER["central_meridian",-165],PARAMETER["scale_factor",0.9996],PARAMETER["false_easting",500000],PARAMETER["false_northing",0],UNIT["metre",1,AUTHORITY["EPSG","9001"]],AXIS["Easting",EAST],AXIS["Northing",NORTH],AUTHORITY["EPSG","32603"]],VERT_CS["EGM2008 height",VERT_DATUM["EGM2008 geoid",2005,AUTHORITY["EPSG","1027"]],UNIT["metre",1,AUTHORITY["EPSG","9001"]],AXIS["Gravity-related height",UP],AUTHORITY["EPSG","3855"]]]
```

**Comment by iannesbitt** [comment]
The use of EGM2008 versus GEOID12B explains most of the extra ~2 m of difference I was seeing between the Cesium terrain and the points. I was using +5.2 as an adjustment but I needed to be using +6.99. So I think I'm comfortable saying this issue is "known" now.

**Comment by iannesbitt** [comment]
After spending a frustrating amount of time trying to find tools to parse `COMPD_CS` CRS objects (neither geodaisy nor PyCRS would recognize valid `COMPD_CS` WKT strings), I've found that (of course) `pyproj` [handles them natively](https://gis.stackexchange.com/a/461622/50005). I think the reason I couldn't find anything when searching for info initially is that they are fairly rarely used and a [relatively new](https://groups.google.com/g/lastools/c/0YPhZce3xIY/m/gxL7ozE-BQAJ) invention of the USGS.

**Comment by iannesbitt** [comment]
See [relevant pyproj docs](https://pyproj4.github.io/pyproj/stable/api/crs/crs.html#pyproj.crs.CRS.sub_crs_list)

**Comment by iannesbitt** [comment]
Similar to [this comment](https://github.com/PermafrostDiscoveryGateway/viz-points/issues/14#issuecomment-1553283957) which describes NGS API for querying GEOIDXX models, here is a query to VDatum for EGM2008:
https://vdatum.noaa.gov/vdatumweb/api/convert?s_x=-165.399282&s_y=64.506181&s_h_frame=WGS84_G1674&s_v_geoid=EGM2008&s_v_frame=EGM2008&t_v_frame=WGS84_G1674&region=ak&t_v_geoid=EGM2008&t_h_frame=WGS84_G1674

I used the following string format to construct this API call:
```python
VDATUM_URL = 'https://vdatum.noaa.gov/vdatumweb/api/convert?s_x=%s&s_y=%s&s_h_frame=%s&s_v_frame=%s&s_v_geoid=%s&t_h_frame=%s&t_v_frame=%s&t_v_geoid=%s&region=%s'
```

I still need to work on converting EPSG codes derived from the WKT to geoid and reference frame names (e.g. `EGM2008`, `WGS84_G1674`)

**Comment by iannesbitt** [comment]
EPSG code / reference frame name extractor and API calls are set up. Now I need to construct a way to extract median X and Y coordinates from LAS files and convert them to lat/lon.

**Comment by iannesbitt** [comment]
Finished a working version of [pyegt](https://iannesbitt.github.io/pyegt) which allows for easy geoid height lookups via one of two NOAA APIs. This will be incorporated into viz-points to be able to convert points ellipsoid-relative height when necessary.

**Comment by iannesbitt** [comment]
This issue is close to being resolved; code is now fully written but issue #23 needs to be resolved in order for testing to take place

**Comment by iannesbitt** [comment]
Tested and working
